code/game/objects/items/weapons/chemical_flamethrower/chemical_flamethrower.dm

```diff
@@ -0,0 +1,231 @@
+			playsound(user, 'sound/weapons/empty.ogg', 100, TRUE)
+			break // Whoops! No ammo!
+
+		if(T == previousturf)
+			continue	// So we don't burn the tile we be standin on
+
+		var/found_obstruction = FALSE
+		for(var/obj/thing in T)
+			if(thing.density && !istype(thing, /obj/structure/table))
+				found_obstruction = TRUE
+				break
+		if(found_obstruction)
+			break
+
+		make_flame(T)
+		get_canister_stats() // In case we ran out of some fuel this fire
+		sleep(1)
+		previousturf = T
+
+/obj/item/chemical_flamethrower/proc/make_flame(turf/spawn_turf)
+	new /obj/effect/fire(spawn_turf, canister_burn_temp, (canister_burn_duration + rand(1, 3) SECONDS), canister_fire_applications) // For that spicy randomness (and to save your ears from all fires extinguishing at the same time)
+
+/*
+  * Uses `amount` ammo from the flamethrower.
+  * Returns `TRUE` if ammo could be consumed, returns `FALSE` if it failed somehow
+  * It will use up ammo if it failed.
+  */
+/obj/item/chemical_flamethrower/proc/use_ammo(amount)
+	var/total_ammo
+	for(var/obj/item/chemical_canister/canister as anything in canisters)
+		total_ammo += canister.ammo
+	if(total_ammo - amount < 0)
+		return FALSE
+
+	var/length = length(canisters)
+	var/difference = amount
+	for(var/i in 0 to length)
+		var/obj/item/chemical_canister/canister = canisters[length - i]
+		if(canister.ammo - difference <= 0)
+			difference -= canister.ammo
+			canister.ammo = 0
+		else
+			canister.ammo -= difference
+			difference = 0
+
+		if(!difference)
+			break
+
+	if(difference)
+		return FALSE
+	return TRUE
+
+/obj/item/chemical_flamethrower/extended
+	name = "extended capacity chemical flamethrower"
+	desc = "A flamethrower that accepts two chemical cartridges to create lasting fires."
+	max_canisters = 2
+
+/obj/item/chemical_canister
+	name = "chemical canister"
+	desc = "A simple canister of fuel. Does not accept any pyrotechnics except for welding fuel."
+	icon = 'icons/obj/tank.dmi'
+	icon_state = "oxygen"
+	container_type = OPENCONTAINER
+	/// How much ammo do we have? Empty at 0.
+	var/ammo = 100
+	/// Which reagent IDs do we accept
+	var/list/accepted_chemicals = list("fuel")
+	/// The burn temperature of our currently stored chemical
+	var/chem_burn_temp = T0C + 300
+	/// The burn duration of our currently stored chemical
+	var/chem_burn_duration = 10 SECONDS
+	/// How many firestacks will our reagent apply
+	var/fire_applications = 1
+	/// The currently stored reagent ID
+	var/current_reagent_id
+	/// How many units of the reagent do we need to have it's effects kick in?
+	var/required_volume = 10
+
+/obj/item/chemical_canister/Initialize(mapload)
+	. = ..()
+	create_reagents(50)
+
+/obj/item/chemical_canister/on_reagent_change()
+	if(!reagents.get_master_reagent_id() || !(reagents.get_master_reagent_id() in accepted_chemicals))
+		reagents.clear_reagents()
+		visible_message("<span class='notice'>[src] doesn't accept the most present chemical!</span>")
```

Comment:
This behavior seems somewhat unintuitive, why nuke all the reagents when you isolate them later?

Reply:
We nuke the reagents because the main reagent wasn't an accepted one
If it is an accepted one, we just isolate it